File: 500-lambda.rb
ZERO = ->(f) { ->(v) {       v   } }
ONE  = ->(f) { ->(v) {    f.(v)  } }
TWO  = ->(f) { ->(v) { f.(f.(v)) } }

TRU = ->(x) { ->(y) { x } }
FLSE = ->(x) { ->(y) { y } }

IS_ZERO = ->(f) { f.(->(v) { FLSE }).(TRU)  }


def to_bool(bool)
  bool.(true).(false)
end

puts to_bool(TRU)
puts to_bool(FLSE)

puts to_bool(IS_ZERO.(ZERO))
puts to_bool(IS_ZERO.(TWO))

# next up - if/else

File: js/500-lambda.js
const ZERO = (f) => { return (v) => { return     v   }}
const ONE  = (f) => { return (v) => { return   f(v)  }}
const TWO  = (f) => { return (v) => { return f(f(v)) }}

const TRUE  = (x) => { return (y) => { return x } }
const FALSE = (x) => { return (y) => { return y } }

const IS_ZERO = (f) => { return f((v) => { return FALSE })(TRUE) }

function toBool(bool) {
  return bool(true)(false)
}

console.log(toBool(TRUE))
console.log(toBool(FALSE))

console.log(toBool(IS_ZERO(ZERO)))
console.log(toBool(IS_ZERO(TWO)))

// next up - if/else
